src/Language/Fixpoint/Solver/PLE.hs

```diff
@@ -266,9 +271,99 @@ withAssms :: InstEnv a -> ICtx -> Diff -> Maybe SubcId -> (InstEnv a -> ICtx ->
-    forM_ assms (SMT.smtAssert ieSMT)
-    act env' ctx' { icAssms = mempty }
+
+  let smtRewrites = if etabeta ieCfg then
+                      concatMap (\i -> extractRewrites $ lookupBindEnv i ieBEnv) delta
+                    else
+                      mempty
+
+  SMT.smtBracket ieSMT "PLE.evaluate" $ do
+    forM_ assms (SMT.smtAssert ieSMT )
+    act env' ctx' { icAssms = mempty, icSMTRewrites = M.union (M.fromList smtRewrites) (icSMTRewrites ctx) }
+  where
+
+    -- [(is$Something foo);
+    --  (~((is$SomethingElse foo));
+    --  ...
+    --  ((prop foo) = 
+    --    (Something v₁ ... vₙ));
+    --  ((prop foo) =
+    --    (Something m₁ ... mₙ));
+    --  ...]
+
+    -- We solve the system v₁ = m₁, v₂ = m₂, ... vₙ = mₙ
+    -- finding equalities between the free variables
+
+    -- Given ((Something v₁ ... vₙ), (Something m₁ ... mₙ))
+    -- or    (v₁,                    (Something m₁ ... mₙ))
+    -- obtain the unification
+    obtainEqualities (e1, e2)
+      | fst1 : args1 <- flatten $ dropECst e1
+      , fst2 : args2 <- flatten $ dropECst e2
+      , EVar cons1 <- dropECst fst1
+      , EVar cons2 <- dropECst fst2
+      , cons1 == cons2
+      , Just _ <- M.lookup cons1 (knDataCtors ieKnowl)
+      = unifyArgumets args1 args2
+    obtainEqualities (e1, e2)
+      | EVar f <- dropECst e1
+      , validVar f
+      , fst2 : _ <- flatten $ dropECst e2
+      , (EVar cons) <- dropECst fst2
+      , Just _ <- M.lookup cons $ knDataCtors ieKnowl
+      = Just [(f, e2)]
+    obtainEqualities (e1, e2)
+      | EVar f <- dropECst e2
+      , validVar f
+      , fst1 : _ <- flatten $ dropECst e1
+      , (EVar cons) <- dropECst fst1
+      , Just _ <- M.lookup cons $ knDataCtors ieKnowl
+      = Just [(f, e1)]
+    obtainEqualities _ = Nothing
+
+    flatten (EApp x xs) = flatten x ++ [ xs ]
+    flatten x = [x]
+
+    unifyArgumets [] [] = Just []
+    unifyArgumets (x:xs) (y:ys)
+      | EVar xn <- dropECst x
+      , validVar xn
+      = fmap ((xn, y) :) (unifyArgumets xs ys)
+    unifyArgumets (x:xs) (y:ys)
+      | EVar yn <- dropECst y
+      , validVar yn
+      = fmap ((yn, x) :) (unifyArgumets xs ys)
+    unifyArgumets (x:xs) (y:ys)
+      | dropECst x == dropECst y
+      = unifyArgumets xs ys
+    unifyArgumets _ _ 
+      = Nothing
+
+    -- Looks inside a list of qualities for equalities 
+    -- that share the lhs and then perform the unification
+    extractRewrites (_, sreft, _)
+      | Reft (_, PAnd ra) <- sr_reft sreft
+      =  concat
+      $  Mb.mapMaybe obtainEqualities
+      $  Mb.mapMaybe (keepTrans . snd)
+      $  Misc.groupList 
+      $  bimap unApply unApply
+     <$> Mb.mapMaybe keepPropEq ra
+    extractRewrites _ = []
+
+    keepPropEq (PAtom Eq e1 e2) = Just (e1, e2)
+    keepPropEq _ = Nothing
+    
+    keepTrans [a, b] = Just (a, b)
+    keepTrans _ = Nothing
+
+-- Find a way to make this check more robust
+-- Check if a symbol is a local variable
+validVar :: Symbol -> Bool
```

Comment:
> ... is requiring local variables on the left of a rewrite (x, e) sufficient to ensure termination? Maybe e should not contain local variables, or some other restriction of the sort.

Here's an example I am imagining:

```Haskell
data C = C (Int -> Int)

{-@ f :: g:(Int -> Int) -> {f:(Int -> Int) | C f = C g && C g = C f} -> { f = g } @-}
f :: (Int -> Int) -> (Int -> Int) -> ()
f _ _ = ()
```

I would expect unification to produce rewrites `(f, g)` and `(g, f)` in this case.

Reply:
This code wont trigger PLE but it can be modified to do so:

```haskell
{-@ LIQUID "--reflection"                @-}
{-@ LIQUID "--ple"                       @-}
{-@ LIQUID "--etabeta"                   @-}
{-@ LIQUID "--save"                      @-}

module Counter where

import Prelude hiding (id, const)

data C = C (Int -> Int)

{-@ reflect id @-}
id :: a -> a
id x = x

{-@ reflect const @-}
const :: a -> b -> a
const x _ = x

{-@ h :: g:_ -> {f:_ | C f = C g && C g = C f} -> {k:_ | k = C f} -> {q:_ | q = C g} -> { f = const id g } @-}
h :: (Int -> Int) -> (Int -> Int) -> C -> C -> ()
h _ _ (C f) (C g) = ()
```

2 things are required to trigger my code:
- Trigger PLE thus we need a non trivial goal for the SMT
- Case splitting to generate the rewrites

It wont cause an infinite loop as the the local variables (f and g) will be rewritten to a third one lets call it (You will get the rewrites `f -> k` and `g -> k`)

Reply:
You can look at the rewrites by adding by adding
```haskell
  traceShowM ("ID" :: String, cidMb)
  traceShowM ("REWRITES: " :: String, pprint smtRewrites)
```
At line 281 in `PLE.hs`

Reply:
The unification is very sensitive to the way in which the equalities are written. Here's our infinite loop.
```Haskell
{-@ LIQUID "--ple" @-}
{-@ LIQUID "--reflection" @-}
{-@ LIQUID "--extensionality" @-}
{-@ LIQUID "--etabeta" @-}

module Test where

data C = C (Int -> Int)

{-@
lemma
  :: k:C
  -> l:C
  -> g:(Int -> Int)
  -> f:(Int -> Int)
  -> { v:()
     |    k = C f
       && k = C (\x:Int -> g x)
       && l = C g
       && l = C (\y:Int -> f y)
     }
  -> { f = g }
@-}
lemma :: C -> C -> (Int -> Int) -> (Int -> Int) -> () -> ()
lemma _ _ _ _ _ = ()
```

Reply:
Um you're right probably we can detect loops by looking at the rewrites but it would also make us lose some rewrites that are ok

Reply:
But probably before we should decide what to do with variables

Reply:
After thinking a bit abou that I underythe issue those rewrites do not come by a case split

Reply:
Deferred by removal of the change.